velox/exec/prefixsort/PrefixSortEncoder.h

```diff
@@ -28,52 +28,177 @@ namespace facebook::velox::exec::prefixsort {
 /// Provides encode/decode methods for PrefixSort.
 class PrefixSortEncoder {
  public:
-  /// 1. Only int64_t is supported now.
-  /// 2. Encoding is compatible with sorting ascending with no nulls.
+  PrefixSortEncoder(bool ascending, bool nullsFirst)
+      : ascending_(ascending), nullsFirst_(nullsFirst){};
+
+  /// Encode native primitive types(such as byte, int16, int32, int64, double,
+  /// float). There will be a separate method to handle string type as
+  /// follow-up.
+  /// 1. The first byte of the encoded result is nullByte, which is
+  ///    determined by the nullsFirst :
+  ///    -If nullsFirst : value is null -> 0, value is not null -> 1
+  ///    -If !nullsFirst : value is null -> 1, value is not null -> 0
+  /// 2. The remaining bytes are the encoding result of value:
+  ///    -If value is null, we set the remaining sizeof(T) bytes to '0', they
+  ///     do not affect the comparison results at all.
+  ///    -If value is not null, we set the remaining bytes as the result of
+  ///     encodeNoNulls, which will consider whether it is ascending or not.
   template <typename T>
-  static FOLLY_ALWAYS_INLINE void encode(T value, char* row);
+  FOLLY_ALWAYS_INLINE void encode(std::optional<T> value, char* dest) const;
+
+  /// 1. Only uint64_t/int64_t/uint32_t/int32_t/float/double is supported now.
+  ///    TODO: Since int16_t/uint16_t do not have build-in swap method, and
+  ///    will add them as a separate PR as follow-up.
+  /// 2. Assuming no nulls.
   template <typename T>
-  static FOLLY_ALWAYS_INLINE void decode(char* row);
+  FOLLY_ALWAYS_INLINE void encodeNoNulls(T value, char* dest) const;
 
  private:
-  FOLLY_ALWAYS_INLINE static uint8_t flipSignBit(uint8_t byte) {
-    return byte ^ 128;
-  }
+  // CompareFlags
+  const bool ascending_;
+  const bool nullsFirst_;
 };
 
-/// Assuming that value is little-endian encoded, we encode it as follows to
-/// make sure encoding is compatible with sorting ascending:
-/// 1 Reverse each byte, for example, 0xaabbccdd becomes 0xddccbbaa.
-/// 2 Flip the sign bit.
-/// The decode logic is exactly the opposite of the above approach.
+template <typename T>
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encode(
```

Comment:
I'm not sure if we have it documented, but the guidelines are the following.

- Minimize the amount of code in header files. Make sure non-trivial method implementations are placed in .cpp file. This helps with compilation times.
- The above is not possible for templates. Hence, use templates only when necessary. Inline templated method if these are not too long. Otherwise, put their implementation into xxx-inl.h file. Optimize for readability.